irohad/execution/command_executor.hpp

```diff
@@ -0,0 +1,333 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_COMMAND_EXECUTOR_HPP
+#define IROHA_COMMAND_EXECUTOR_HPP
+
+#include <boost/format.hpp>
+#include <boost/variant/static_visitor.hpp>
+
+#include "ametsuchi/wsv_command.hpp"
+#include "ametsuchi/wsv_query.hpp"
+#include "builders/default_builders.hpp"
+#include "common/result.hpp"
+#include "interfaces/commands/add_asset_quantity.hpp"
+#include "interfaces/commands/add_peer.hpp"
+#include "interfaces/commands/add_signatory.hpp"
+#include "interfaces/commands/append_role.hpp"
+#include "interfaces/commands/create_account.hpp"
+#include "interfaces/commands/create_asset.hpp"
+#include "interfaces/commands/create_domain.hpp"
+#include "interfaces/commands/create_role.hpp"
+#include "interfaces/commands/detach_role.hpp"
+#include "interfaces/commands/grant_permission.hpp"
+#include "interfaces/commands/remove_signatory.hpp"
+#include "interfaces/commands/revoke_permission.hpp"
+#include "interfaces/commands/set_account_detail.hpp"
+#include "interfaces/commands/set_quorum.hpp"
+#include "interfaces/commands/subtract_asset_quantity.hpp"
+#include "interfaces/commands/transfer_asset.hpp"
+
+namespace iroha {
+
+  /**
+   * Error for command execution.
+   * Contains command name, as well as an error message
+   */
+  struct ExecutionError {
+    std::string command_name;
+    std::string error_message;
+
+    std::string toString() const {
+      return (boost::format("%s: %s") % command_name % error_message).str();
+    }
+  };
+
+  /**
+   * ExecutionResult is a return type of all execute functions.
+   * If execute is successful, result will not contain anything (void value),
+   * because execute does not return any value.
+   * If execution is not successful, ExecutionResult will contain Execution
+   * error with explanation
+   *
+   * Result is used because it allows to clear distinction between two states
+   * - value and error. If we just returned error, it would be confusing, what
+   * value of an error to consider as a successful state of execution
+   */
+  using ExecutionResult = iroha::expected::Result<void, ExecutionError>;
+
+  class CommandExecutor : public boost::static_visitor<ExecutionResult> {
+   public:
+    CommandExecutor(std::shared_ptr<iroha::ametsuchi::WsvQuery> queries,
+                    std::shared_ptr<iroha::ametsuchi::WsvCommand> commands);
+
+    ExecutionResult operator()(
+        const shared_model::detail::PolymorphicWrapper<
```

Comment:
Consider making a variant of const refs, such as `variant<const shared_model::interface::AddAssetQuantity&, const shared_model::interface:: AddPeer&, ...>`, and a mapping visitor, which will take variant of wrappers, and return the const ref variant. The const ref variant and types can be used in this visitor.

Reply:
We decided to remove it in https://soramitsu.atlassian.net/secure/RapidBoard.jspa?rapidView=9&view=planning&selectedIssue=IR-1011
@victordrobny  please, leave TODO with IR-1011 to fix that later